crowdaction/main.go

```diff
@@ -0,0 +1,294 @@
+	//could use a parameter store here
+	//example: current date = 2021.12.01 => retrieve up to and including crowdaction end date=2021.12.30
+	dateLimit = 30
+
+	//access pattern getCrowdaction
+	//item has PK="act#<crowdactionID>" and SK="act#<crowdactionID>"
+	prefixPKcrowdactionID = "act#"
+	prefixSKcrowdactionID = "act#"
+
+	//access pattern getActiveCrowdactions
+	//item has PK="act_end#date_end" and SK="act#<crowdactionID>"
+	prefixPKcrowdaction_date_end = "act_end#"
+
+	//access pattern getEligibleToJoinCrowdactions
+	//item has PK="act_join#date_limit_join" and SK="date_start#act#<crowdactionID>"
+	prefixPKcrowdaction_date_limit_join = "act_join#"
+)
+
+type pkError struct {
+	Pk  string `json:"pk"`
+	Err string `json:"error"`
+}
+
+var (
+	tableName = os.Getenv("CROWDACTION_TABLE")
+	dbClient  *dynamodb.DynamoDB
+)
+
+////get list of crowd actions
+func getListCrowdaction(req events.APIGatewayV2HTTPRequest, status string) (events.APIGatewayProxyResponse, error) {
+
+	var pk, sk string
+	var items []*dynamodb.QueryOutput
+	var crowdaction Crowdaction
+	var errors []string
+
+	mapErr := make(map[string]error)
+
+	//get current date
+	dateCurrent := time.Now()
+
+	for i := 0; i < dateLimit; i++ {
+
+		date := dateCurrent.AddDate(0, 0, i).Format(dateFormat)
+		switch status {
+		case "active":
+			pk = prefixPKcrowdaction_date_end + date
+			sk = ""
+		case "joinable":
+			pk = prefixPKcrowdaction_date_limit_join + date
+			//add one day so the result includes the crowdactions starting on the current day as well
+			sk = dateCurrent.AddDate(0, 0, 1).Format(dateFormat)
+		}
+		//get items for a partition key
+		result, err := getItems(pk, sk)
+		//error handling https://play.golang.org/p/D5xTeZP9VnU
+		if err != nil {
+			mapErr[pk] = err
+		}
+		if result != nil {
+			//add partition key result to the total result for the date range
+			items = append(items, result)
+		}
+	}
+
+	if len(mapErr) > 0 {
+		e := pkError{}
+		for k, v := range mapErr {
+			e.Pk = k
+			e.Err = v.Error()
+			jsonErr, err := json.Marshal(e)
+			if err != nil {
+				errors = append(errors, err.Error())
+			} else {
+				errors = append(errors, string(jsonErr))
+			}
+		}
+
+		body := "[" + strings.Join(errors, ",") + "]"
+		return events.APIGatewayProxyResponse{Body: body, StatusCode: http.StatusInternalServerError}, nil
+
+	}
+
+	if items == nil {
+		body, _ := json.Marshal(map[string]string{"message": "no active crowdactions found"})
+		return events.APIGatewayProxyResponse{
+			Body:       string(body),
+			StatusCode: http.StatusNotFound,
+		}, nil
+	}
+
+	//no error yet if we are here
+	var listCrowdaction []string
+	for _, v := range items {
+		for _, item := range v.Items {
+			err := dynamodbattribute.UnmarshalMap(item, &crowdaction)
+			if err != nil {
+				return events.APIGatewayProxyResponse{Body: err.Error(), StatusCode: http.StatusBadRequest}, nil
+			}
+			action, err := json.Marshal(map[string]interface{}{"data": crowdaction})
```

Comment:
Not all fields of the crowdaction should be returned!
(Exclude `Crowdaction.Code`)

Reply:
Don't they need it to display whether the action could only be joined by password?

Reply:
Good point, but then we should still not return the password.
In this case, we should maybe have an additional field `RequiresPassword`